src/Evolution/Systems/GrMhd/ValenciaDivClean/KastaunEtAl.cpp

```diff
@@ -0,0 +1,342 @@
+        equation_of_state_.specific_internal_energy_lower_bound(rho_hat),
+        equation_of_state_.specific_internal_energy_upper_bound(rho_hat));
+    // Pressure from EOS
+    double p_hat = std::numeric_limits<double>::signaling_NaN();
+    if constexpr (ThermodynamicDim == 1) {
+      p_hat = get(
+          equation_of_state_.pressure_from_density(Scalar<double>(rho_hat)));
+    } else if constexpr (ThermodynamicDim == 2) {
+      p_hat = get(equation_of_state_.pressure_from_density_and_energy(
+          Scalar<double>(rho_hat), Scalar<double>(epsilon_hat)));
+    }
+    return Primitives{rho_hat, w_hat, p_hat, epsilon_hat, q_bar, r_bar_squared};
+  }
+
+  double operator()(const double mu) const noexcept {
+    const auto [rho_hat, w_hat, p_hat, epsilon_hat, q_bar, r_bar_squared] =
+        primitives(mu);
+    // Equation (43)
+    const double a_hat = p_hat / (rho_hat * (1.0 + epsilon_hat));
+    const double h_hat = (1.0 + epsilon_hat) * (1.0 + a_hat);
+    // Equations (46) - (48)
+    const double nu_hat = std::max(
+        h_hat / w_hat, (1.0 + a_hat) * (1.0 + q_bar - mu * r_bar_squared));
+    // Equations (44) - (45)
+    return mu - 1.0 / (nu_hat + mu * r_bar_squared);
+  }
+
+ private:
+  const double q_;
+  const double r_squared_;
+  const double b_squared_;
+  const double r_dot_b_squared_;
+  const double rest_mass_density_times_lorentz_factor_;
+  const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+      equation_of_state_;
+  const double h_0_;
+  const double v_0_squared_;
+};
+}  // namespace
+
+template <size_t ThermodynamicDim>
+std::optional<PrimitiveRecoveryData> KastaunEtAl::apply(
+    const double /*initial_guess_pressure*/, const double total_energy_density,
+    const double momentum_density_squared,
+    const double momentum_density_dot_magnetic_field,
+    const double magnetic_field_squared,
+    const double rest_mass_density_times_lorentz_factor,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+        equation_of_state) noexcept {
+  // Master function see Equation (44)
+  const auto f_of_mu =
+      FunctionOfMu<ThermodynamicDim>{total_energy_density,
+                                     momentum_density_squared,
+                                     momentum_density_dot_magnetic_field,
+                                     magnetic_field_squared,
+                                     rest_mass_density_times_lorentz_factor,
+                                     equation_of_state};
+
+  // mu is 1 / (h W) see Equation (26)
+  double one_over_specific_enthalpy_times_lorentz_factor =
+      std::numeric_limits<double>::signaling_NaN();
+  try {
+    // Bracket for master function, see Sec. II.F
+    const auto [lower_bound, upper_bound] = f_of_mu.root_bracket(
+        rest_mass_density_times_lorentz_factor, absolute_tolerance_,
+        relative_tolerance_, max_iterations_);
+
+    one_over_specific_enthalpy_times_lorentz_factor =
```

Comment:
I'd find it helpful to add a comment `// Try to recover primitives` because initially I was confused that there's a second root find in bracketing the master function. That is, I didn't know where comment `296` ended